Source/WebKit/NetworkProcess/haiku/NetworkDataTaskHaiku.cpp

```diff
@@ -135,33 +166,168 @@ NetworkDataTask::State NetworkDataTaskHaiku::state() const
+            // AuthenticationNeeded may have aborted the request
+            // so we need to make sure we can continue.
+            
+            if (m_currentRequest.isNull())
+                return;
+        }
+    }
+
+    if (!m_client)
+        return;
+
+    if (m_redirected) {
+        m_redirectionTries--;
+
+        if (m_redirectionTries == 0) {
+            ResourceError error(url.host().utf8().data(), 400, url,
+                "Redirection limit reached");
+            
+            m_networkLoadMetrics.responseEnd = MonotonicTime::now() - m_startTime;
+            m_networkLoadMetrics.markComplete();
+            m_client->didCompleteWithError(error,m_networkLoadMetrics);
+            return;
+        }
+
+        // Notify the client that we are redirecting.
+        ResourceRequest request = m_currentRequest;
+        ResourceResponse responseCopy = response;
+        request.setURL(url);
+    	fprintf(stderr,"((((((( redirecting to %s)))))))\n",request.url().string().utf8().data());
+    	m_client->willPerformHTTPRedirection(WTFMove(responseCopy),WTFMove(request), [this](const ResourceRequest& newRequest){
+    		
+	    	if(newRequest.isNull() || m_state == State::Canceling)
+	    	return;
+	    	
+	    	m_startTime = MonotonicTime::now();//network metrics
+	    	
+	    	if( m_state != State::Suspended ){
+	    		m_state = State::Suspended;
+	    		resume();
+	    	}
+	    });
+    	
+    } else {
+        ResourceResponse responseCopy = response;
+        m_client->didReceiveResponse(WTFMove(responseCopy),[this](WebCore::PolicyAction policyAction){
+        	if(m_state == State::Canceling || m_state == State::Completed){
+        		return;
+        	}
+        	});
+    }
 }
 void NetworkDataTaskHaiku::DataReceived(BUrlRequest* caller, const char* data, off_t position,
         ssize_t size)
 {
-	if(size < 0 )
-	{
-		return;
-	}
-	m_client->didReceiveData(SharedBuffer::create(data,size));
+	if (m_currentRequest.isNull())
+        return;
+
+    if (!m_client)
+        return;
+
+    // don't emit the "Document has moved here" type of HTML
+    if (m_redirected)
+        return;
+
+    if (position != m_position)
+    {
+        debugger("bad redirect");
+        return;
+    }
+
+    if (size > 0) {
+        m_responseDataSent = true;
+        
+        
+        runOnMainThread([this,data = data, size = size]{
+			m_client->didReceiveData(SharedBuffer::create(data,size));
+		});
+    }
+
+    m_position += size;
 }
 void NetworkDataTaskHaiku::UploadProgress(BUrlRequest* caller, ssize_t bytesSent, ssize_t bytesTotal)
 {
 	
 }
 void NetworkDataTaskHaiku::RequestCompleted(BUrlRequest* caller, bool success)
 {
-	
+
 }
 bool NetworkDataTaskHaiku::CertificateVerificationFailed(BUrlRequest* caller, BCertificate& certificate, const char* message)
 {
-	
+
+}
+void NetworkDataTaskHaiku::DebugMessage(BUrlRequest* caller,BUrlProtocolDebugMessage type,const char* text)
+{
```

Comment:
use BeDC here to log the request progress?

Reply:
Will do it as i was focusing on reworking minibrowser